File: javascript/allConstruct.js
function allConstruct(target, wordBank) {
    if (target.length === 0) return [[]]

    let allWays = []

    for (let word of wordBank) {
        if (target.startsWith(word)) {
            const suffixWays = allConstruct(target.substring(word.length), wordBank)
            const targetWays = suffixWays.map(way => [word, ...way])
            allWays.push(...targetWays)
        }
    }

    return allWays
}

console.log(allConstruct("purple", ["purp", "p", "ur", "le", "purpl"]))
console.log(allConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd", "ef", "c"]))
console.log(allConstruct("skateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar", "skate"]))
console.log(allConstruct("aaaaaaaaaaaaaaaaaaaaaaaaaz", ["a", "aa", "aaa", "aaaa", "aaaaa", "aaaaaa", "aaaaaaa", "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa"]))

File: kotlin/allConstruct.kt
fun allConstruct(target: String, wordBank: Array<String>): List<List<String>> {
    if (target.isEmpty()) return listOf(listOf())

    val allWays: MutableList<List<String>> = mutableListOf()

    wordBank.forEach { word ->
        if (target.startsWith(word)) {
            val remainingWays: List<List<String>> = allConstruct(target.slice(word.length..(target.length - 1)), wordBank);
            val targetWays: List<List<String>> = remainingWays.map() { way -> listOf(word) + way }
            targetWays.forEach(allWays::add)
        }
    }

    return allWays
}

fun main(): Unit {
    println(allConstruct("purple", arrayOf("purp", "p", "ur", "le", "purpl")))
    println(allConstruct("strawberry", arrayOf("straw", "st", "ra", "w", "ber", "berry")))
    println(allConstruct("yellow", arrayOf("y", "ll", "o", "ow", "w", "e", "llow", "yel", "low", "ellow")))
    println(allConstruct("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaf", arrayOf("a", "aa", "aaa", "aaaa", "aaaaa", "aaaaaa", "aaaaaaa", "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa")))
}